### PR TITLE
[dattri.task] Add `loss_func` and make `target_func` optional

### PR DESCRIPTION
**Pull request by TheaperDeng**
### 1. Motivation and Context
We want to make 2 concepts clearer.

- `loss_func`: the loss function used to train the model to be attributed. Normally, this one is used to estimate the changes of parameter when one data point is up/down weighted (IF style) or between each iterations (TracIN style).
- `target_func`: the target function we want to attribute on. This one is normally set to be the same as loss function but not necessarily to be the same. so we make it optional in `dattri.task.AttributionTask`.

### 2. Summary of the change
1. add `loss_func` in `dattri.task.AttributionTask` and make `target_func` optional.
2. change some internal implementation of `dattri.algorithm` (no API change)

### 3. What tests have been added/updated for the change?
- [ ] Unit test: Typically, this should be included if you implemented a new function/fixed a bug.

### CODE REVIEW

**Comment by tingwl0122** [comment]
LGTM now

**Review by tingwl0122 — APPROVED** [review]
